misc/graphify.js

```diff
@@ -0,0 +1,124 @@
+var _ = require('lodash');
+var fs = require('fs');
+var geoip = require('geoip-lite');
+var crawler = require('../src/crawler.js');
+var normalizeIpp = crawler.normalizeIpp;
+var normalizePubKey = crawler.normalizePubKey;
+
+var rawCrawl = require('../' + process.argv.slice(2)[0]);
+
+function graphify(rawCrawl) {
+
+  var out = {nodes: [], links: [], iplinks: []}
+
+  // storage
+  var nodes = []
+  var links = {}
+  var indices = {}
+  var keyToIp = {}
+
+  // Collect all unique nodes and save their indices
+  for (var ipp in rawCrawl.data) {
+    indices[ipp] = nodes.length
+    if (nodes.length == 0) {
+      nodes.push({ipp: ipp, connections: 0, entry: 1})
+    } else {
+      nodes.push({ipp: ipp, connections: 0, entry: 0})
+    }
+    keyToIp[ipp]
+  }
+
+  // For each unique node
+  for(i in nodes) {
+    node = nodes[i];
+    ipp = node.ipp;
+    peers = rawCrawl.data[ipp].overlay.active;
+
+    // For each (not necessarily unique) peer
+    for (p_index in peers) {
+      peer = peers[p_index];
+
+      // Get ipp
+      try {
+        peer_ipp = normalizeIpp(peer.ip, peer.port);
+      } catch (err) {
+        peer_ipp = undefined;
+      }
+      peer_i = indices[peer_ipp];
+
+      // If it has an ipp
+      if (peer_ipp && peer_i !== undefined) {
+        peer_pk = normalizePubKey(peer.public_key);
+        peer_v = peer.version;
+        peer_t = peer.type;
+
+        // Fill out info in unique node list
+        peer_node = nodes[peer_i];
+
+        //console.error(peer_ipp)
+        if (!peer_node.pk) {
+          peer_node.pk = peer_pk
+        }
+        if (!peer_node.v) {
+          peer_node.v = peer_v
+        }
+        if (!peer_node.loc) {
+          geoloc = geoip.lookup(peer_ipp.split(':')[0])
+          peer_node.loc = geoloc.country + '_' + geoloc.city
+        }
+
+        // Make link
+        var link = undefined
+        if (peer_t) {
+          // Get link
+          if (peer_t == "in") {
+            link = [i, peer_i]
+          }
+          else if (peer_t == "out") {
+            link = [peer_i, i]
+          }
+          else if (peer_t == "peer") {
+            if (peer.ip) {
+              if (peer.ip.split(":").length == 2)
+                link = [peer_i, i]
+              else
+                link = [i, peer_i]
+            }
+          }
+
+          if (peer_t && !links[peer_t]) {
```

Comment:
I don't understand. The type is only `in` or `out`. It looks like we only write the 2 item array to `links`.